constraint/pkg/client/drivers/local/driver.go

```diff
@@ -287,25 +296,45 @@ func (d *Driver) Query(ctx context.Context, target string, constraints []*unstru
 
 		kindResults, err := drivers.ToResults(constraintsMap, resultSet)
 		if err != nil {
-			return nil, nil, err
-		}
-
-		for _, result := range kindResults {
-			result.EvaluationMeta = RegoEvaluationMeta{
-				TemplateRunTime: float64(evalEndTime.Nanoseconds()) / 1000000,
-				ConstraintCount: uint(len(kindResults)),
-			}
+			return &drivers.QueryResponse{Err: err}
 		}
 
 		results = append(results, kindResults...)
+
+		if d.gatherMetrics {
+			// for the batch of evaluated constraints, report their evaluation metadata
+			evalRunTime := float64(evalEndTime.Nanoseconds()) / 1000000
+
+			statsEntries = append(statsEntries,
+				&instrumentation.StatsEntry{
+					Key: kind,
+					Stats: []*instrumentation.Stat{
+						{
+							Name:        templateRunTimeMillisName,
+							Value:       &instrumentation.Float64Value{ConcreteValue: evalRunTime},
+							Description: templateRunTimeMillisDescription,
```

Comment:
How do we know which constraint template this metric is associated with?

Reply:
the general shape of `StatsEntry` is 

```
Key interface{}
Stats []*Stats
```

`Key ` here is specified on line 310 and it refers to the constraint kind evaluated. Here's what a `StatsEntry` would look like marshalled

```
    "key": "K8sUniqueServiceSelector",
    "stats": [
      {
        "name": "templateRunTimeMillis",
        "value": 0.37196,
        "description": "the number of milliseconds it took to evaluate all constraints for a template"
      },
      {
        "name": "engineType",
        "value": "rego",
        "description": "the underlying engine that performed the evaluation"
      },
      {
        "name": "constraintCount",
        "value": 1,
        "description": "the number of constraints that were evaluated at the same time for the given constraint kind"
      }
    ]
```

Reply:
I see, how do users know how to interpret "key"... is it the name of the template? The constraint?

Also, "engine" sticks out a bit here, these metrics seem like they'll lose coherence if they get remixed or combined with other primitives.

Reply:
> "key"

from some of our chats, I renamed `key` to `scope` but I think your question still stands!

> ... is it the name of the template? The constraint?

in the past we used `ConstraintKey` type as the type. To my mind, a consumer could type check in a `switch` and handle the key types as appropriate. However, using the template name (of type string) now I can see how a consumer won't be able to know what the `string` type stands for now.

Should I add another string to the `StatsEntry` like `ScopeType` that states what the Scope is actually? Something like:

```
{
"Scope": "K8sUniqueServiceSelector",
"ScopeType": "TemplateName",
"Stats": [...]
}
``` 

> Also, "engine" sticks out a bit here, these metrics seem like they'll lose coherence if they get remixed or combined with other primitives.

can you please offer an example here? I have a few thoughts but I'd like to understand scenarios where the `engineType` is problematic before sharing.

Reply:
we've added `scope` to denote the "what kind of" and `statsFor` to denote which specifically of the kind the stats refer to.